{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6bb0f49a",
   "metadata": {},
   "source": [
    "# Analyze CHIKV Passaging Mutations\n",
    "\n",
    "This notebook converts the nucleotide numbering of the passaging mutations from [this](https://www.nature.com/articles/s41598-018-34561-x#additional-information) paper into the same reference amino acid numbering we used for our DMS library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "f4f8332e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "80bd29ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: EU224268.1\n",
      "Name: EU224268\n",
      "Description: Cloning vector pCHIKV-LR ic, complete sequence\n",
      "Number of features: 8\n",
      "/molecule_type=DNA\n",
      "/topology=circular\n",
      "/data_file_division=SYN\n",
      "/date=24-NOV-2007\n",
      "/accessions=['EU224268']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Cloning vector pCHIKV-LR ic\n",
      "/organism=Cloning vector pCHIKV-LR ic\n",
      "/taxonomy=['other sequences', 'artificial sequences', 'vectors']\n",
      "/references=[Reference(title='Infectious clones of Chikungunya virus (La Reunion isolate) for vector competence studies', ...), Reference(title='Direct Submission', ...)]\n",
      "Seq('ATGGCTGCGTGAGACACACGTAGCCTACCAGTTTCTTACTGCTCTACTCTGCAA...TAG')\n"
     ]
    }
   ],
   "source": [
    "# Get the genbank record for the strain used in the study\n",
    "Entrez.email = \"email@email.com\"\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=\"EU224268\", rettype=\"gb\", retmode=\"text\")\n",
    "strain = SeqIO.read(handle, \"genbank\")\n",
    "handle.close()\n",
    "print(strain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "423b42c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the start of the structural polyprotein gene (E glycoprotein)\n",
    "for feature in strain.features:\n",
    "    if feature.type == \"CDS\":\n",
    "        if feature.qualifiers.get(\"product\")[0] == 'structural polyprotein':\n",
    "            # Use the 1-indexed start position\n",
    "            gene_start = feature.location.start + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "6366224f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>Passage</th>\n",
       "      <th>Region</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AA_REF</th>\n",
       "      <th>AA_ALT</th>\n",
       "      <th>FREQ</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Species</th>\n",
       "      <th>AA_POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8549</td>\n",
       "      <td>51</td>\n",
       "      <td>E2</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.04</td>\n",
       "      <td>A20-1</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8549</td>\n",
       "      <td>15</td>\n",
       "      <td>E2</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.25</td>\n",
       "      <td>U4-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8564</td>\n",
       "      <td>15</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>0.03</td>\n",
       "      <td>U4-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8566</td>\n",
       "      <td>50</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.48</td>\n",
       "      <td>C6-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8566</td>\n",
       "      <td>25</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.25</td>\n",
       "      <td>C6-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS  Passage Region REF ALT AA_REF AA_ALT  FREQ Sample     Species  AA_POS\n",
       "0  8549       51     E2   A   G      K      R  0.04  A20-1     aegypti     328\n",
       "1  8549       15     E2   A   G      K      R  0.25   U4-1  albopictus     328\n",
       "2  8564       15     E2   T   C      V      A  0.03   U4-1  albopictus     333\n",
       "3  8566       50     E2   T   C      Y      H  0.48   C6-1  albopictus     334\n",
       "4  8566       25     E2   T   C      Y      H  0.25   C6-1  albopictus     334"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the passaging mutations and get the protein position in the polyprotein\n",
    "passaging_mutations = pd.read_csv('mutations.csv')\n",
    "passaging_mutations['AA_POS'] = ((passaging_mutations['POS'] - gene_start) // 3) + 1\n",
    "passaging_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "632cdc90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>Passage</th>\n",
       "      <th>Region</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AA_REF</th>\n",
       "      <th>AA_ALT</th>\n",
       "      <th>FREQ</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Species</th>\n",
       "      <th>AA_POS</th>\n",
       "      <th>sequential_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8549</td>\n",
       "      <td>51</td>\n",
       "      <td>E2</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.04</td>\n",
       "      <td>A20-1</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>328</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8549</td>\n",
       "      <td>15</td>\n",
       "      <td>E2</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.25</td>\n",
       "      <td>U4-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>328</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8564</td>\n",
       "      <td>15</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>0.03</td>\n",
       "      <td>U4-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>333</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8566</td>\n",
       "      <td>50</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.48</td>\n",
       "      <td>C6-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>334</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8566</td>\n",
       "      <td>25</td>\n",
       "      <td>E2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.25</td>\n",
       "      <td>C6-1</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>334</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS  Passage Region REF ALT AA_REF AA_ALT  FREQ Sample     Species  \\\n",
       "0  8549       51     E2   A   G      K      R  0.04  A20-1     aegypti   \n",
       "1  8549       15     E2   A   G      K      R  0.25   U4-1  albopictus   \n",
       "2  8564       15     E2   T   C      V      A  0.03   U4-1  albopictus   \n",
       "3  8566       50     E2   T   C      Y      H  0.48   C6-1  albopictus   \n",
       "4  8566       25     E2   T   C      Y      H  0.25   C6-1  albopictus   \n",
       "\n",
       "   AA_POS  sequential_site  \n",
       "0     328               68  \n",
       "1     328               68  \n",
       "2     333               73  \n",
       "3     334               74  \n",
       "4     334               74  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the AA_POS into the sequential position relative to our DMS data\n",
    "passaging_mutations['sequential_site'] = ((passaging_mutations['AA_POS'] - 262 + 1) + 1)\n",
    "passaging_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "3faed8e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_site</th>\n",
       "      <th>AA_REF</th>\n",
       "      <th>AA_ALT</th>\n",
       "      <th>FREQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>Passage</th>\n",
       "      <th>Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3(E2)</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.04</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>51</td>\n",
       "      <td>A20-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3(E2)</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0.25</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>15</td>\n",
       "      <td>U4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>0.03</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>15</td>\n",
       "      <td>U4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9(E2)</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.48</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>50</td>\n",
       "      <td>C6-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9(E2)</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0.25</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>25</td>\n",
       "      <td>C6-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reference_site AA_REF AA_ALT  FREQ     Species  Passage Sample\n",
       "0          3(E2)      K      R  0.04     aegypti       51  A20-1\n",
       "1          3(E2)      K      R  0.25  albopictus       15   U4-1\n",
       "2          8(E2)      V      A  0.03  albopictus       15   U4-1\n",
       "3          9(E2)      Y      H  0.48  albopictus       50   C6-1\n",
       "4          9(E2)      Y      H  0.25  albopictus       25   C6-1"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the position from the DMS sitemap to the passaging mutations data\n",
    "sitemap = pd.read_csv('../../../data/site_numbering_map.csv')\n",
    "\n",
    "# Join with the passaging mutations by the sequential_site\n",
    "passaging_mutations = passaging_mutations.merge(sitemap, on='sequential_site', how='left')\n",
    "columns_to_keep = ['reference_site', 'AA_REF', 'AA_ALT', 'FREQ', 'Species', 'Passage', 'Sample']\n",
    "passaging_mutations = passaging_mutations[columns_to_keep]\n",
    "passaging_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d4bbdabe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to a new CSV file\n",
    "passaging_mutations.to_csv('passaging_mutations.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextstrain",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
