{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66ee1de3",
   "metadata": {},
   "source": [
    "# Filter `ivar` variants\n",
    "\n",
    "I used `ivar` to identify variants relative to the sequence of the DMS library strain [MW473668](https://www.ncbi.nlm.nih.gov/nuccore/MW473668). In this notebook, I filter the variants to get only the coding mutations in the CHIKV E glycoprotein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "8505eb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "35dee6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_variants = pd.read_csv(\"../../results/variants/variants.ivar.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e9ba11",
   "metadata": {},
   "source": [
    "Only keep the variants with a more than 100 reads at a site, a minimum frequency of 5%, a significant p-value < 0.05 and variants in the 'structural polyprotein' that encode the CHIKV E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "db6101bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the variants\n",
    "min_depth = 100\n",
    "min_freq = 0.05\n",
    "pvalue = 0.05\n",
    "feature = \"structural polyprotein:\"\n",
    "filtered_variants = raw_variants[\n",
    "    (raw_variants['GFF_FEATURE'] == feature) &\n",
    "    (raw_variants['TOTAL_DP'] >= min_depth) &\n",
    "    (raw_variants['ALT_FREQ'] >= min_freq) &\n",
    "    (raw_variants['PVAL'] <= pvalue)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bbbf301",
   "metadata": {},
   "source": [
    "Convert the numbering so it's relative to each region in the CHIKV E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "de95b972",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regions in the polyprotein\n",
    "polyprotein_numbering = {\n",
    "    \"C\": [1, 261],\n",
    "    \"E3\": [262, 325],\n",
    "    \"E2\": [326, 748],\n",
    "    \"6K\": [749, 809],\n",
    "    \"E1\": [810, 1248],\n",
    "}\n",
    "# Map the mature peptide region and position for each variant\n",
    "def map_peptide_and_position(pos_aa, polyprotein_numbering):\n",
    "    if pd.isna(pos_aa):\n",
    "        return None, None\n",
    "    \n",
    "    for peptide, (start, end) in polyprotein_numbering.items():\n",
    "        if start <= pos_aa <= end:\n",
    "            relative_pos = pos_aa - start + 1  # +1 for 1-based indexing\n",
    "            return peptide, relative_pos\n",
    "    \n",
    "    # If position doesn't fall within any range\n",
    "    return None, None\n",
    "# Apply the function to create both columns\n",
    "filtered_variants = filtered_variants.copy()\n",
    "filtered_variants[['peptide', 'peptide_position']] = filtered_variants['POS_AA'].apply(\n",
    "    lambda x: pd.Series(map_peptide_and_position(x, polyprotein_numbering))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b0fe17f",
   "metadata": {},
   "source": [
    "Only keep non-synonymous mutations in the subunits of the E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "4c48aa58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "      <th>Accession</th>\n",
       "      <th>region</th>\n",
       "      <th>region_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>52</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>52</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>104</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>7122</td>\n",
       "      <td>3359</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>6K</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>9845</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1717</td>\n",
       "      <td>939</td>\n",
       "      <td>5524</td>\n",
       "      <td>2489</td>\n",
       "      <td>0.755884</td>\n",
       "      <td>7308.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>6K</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      POS REF ALT  REF_DP  REF_RV  ALT_DP  ALT_RV  frequency   depth  \\\n",
       "270  9618   T   C       0       0     185      52   1.000000   185.0   \n",
       "271  9626   G   T       1       0     190      52   0.994764   191.0   \n",
       "272  9644   G   A       0       0     373     104   0.997326   374.0   \n",
       "276  9842   C   A       8       3    7122    3359   0.997898  7137.0   \n",
       "277  9845   C   T    1717     939    5524    2489   0.755884  7308.0   \n",
       "\n",
       "           p_value REF_CODON REF_AA ALT_CODON ALT_AA   Accession region  \\\n",
       "270  5.638590e-131       GTG      V       GCG      A  SRR7613172     E2   \n",
       "271  1.848620e-132       GTG      V       TTG      L  SRR7613172     E2   \n",
       "272  1.421440e-260       GTA      V       ATA      I  SRR7613172     E2   \n",
       "276   0.000000e+00       CTG      L       ATG      M  SRR7613172     6K   \n",
       "277   0.000000e+00       CAA      Q       TAA      *  SRR7613172     6K   \n",
       "\n",
       "     region_position  \n",
       "270              368  \n",
       "271              371  \n",
       "272              377  \n",
       "276               20  \n",
       "277               21  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop some columns from the dataframe\n",
    "columns_to_drop = [\"REGION\", \"GFF_FEATURE\", \"POS_AA\", \"PASS\", \"ALT_QUAL\", \"REF_QUAL\"]\n",
    "filtered_variants.drop(columns=columns_to_drop, inplace=True)\n",
    "# Rename columns for clarity\n",
    "filtered_variants.rename(columns={\n",
    "    \"ALT_FREQ\": \"frequency\",\n",
    "    \"TOTAL_DP\": \"depth\",\n",
    "    \"PVAL\": \"p_value\",\n",
    "    \"peptide\": \"region\",\n",
    "    \"peptide_position\": \"region_position\"\n",
    "}, inplace=True)\n",
    "# Convert region_position to integer\n",
    "filtered_variants['region_position'] = filtered_variants['region_position'].astype('Int64')\n",
    "# Filter to keep only regions in the region list\n",
    "regions_to_keep = [\"E1\", \"E2\", \"6K\", \"E3\"]\n",
    "filtered_variants = filtered_variants[\n",
    "    filtered_variants['region'].isin(regions_to_keep)\n",
    "]\n",
    "# Only keep non-synonymous variants where REF_AA != ALT_AA\n",
    "filtered_variants = filtered_variants[\n",
    "    filtered_variants['REF_AA'] != filtered_variants['ALT_AA']\n",
    "]\n",
    "filtered_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfbcef4c",
   "metadata": {},
   "source": [
    "Get the 'fixed' variants. These are variants with a frequency that's greater than `1 - minimum frequency` such that they cannot be converted to minor alleles with a frequency > our `minimum frequency`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "03c34181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "      <th>Accession</th>\n",
       "      <th>region</th>\n",
       "      <th>region_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>52</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>52</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>104</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>E2</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>7122</td>\n",
       "      <td>3359</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>6K</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9909</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>28</td>\n",
       "      <td>13</td>\n",
       "      <td>8577</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.995705</td>\n",
       "      <td>8614.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>SRR7613172</td>\n",
       "      <td>6K</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  REF_RV  ALT_DP  ALT_RV  frequency   depth  \\\n",
       "0  9618   T   C       0       0     185      52   1.000000   185.0   \n",
       "1  9626   G   T       1       0     190      52   0.994764   191.0   \n",
       "2  9644   G   A       0       0     373     104   0.997326   374.0   \n",
       "3  9842   C   A       8       3    7122    3359   0.997898  7137.0   \n",
       "4  9909   T   G      28      13    8577    4230   0.995705  8614.0   \n",
       "\n",
       "         p_value REF_CODON REF_AA ALT_CODON ALT_AA   Accession region  \\\n",
       "0  5.638590e-131       GTG      V       GCG      A  SRR7613172     E2   \n",
       "1  1.848620e-132       GTG      V       TTG      L  SRR7613172     E2   \n",
       "2  1.421440e-260       GTA      V       ATA      I  SRR7613172     E2   \n",
       "3   0.000000e+00       CTG      L       ATG      M  SRR7613172     6K   \n",
       "4   0.000000e+00       TTT      F       TGT      C  SRR7613172     6K   \n",
       "\n",
       "   region_position  \n",
       "0              368  \n",
       "1              371  \n",
       "2              377  \n",
       "3               20  \n",
       "4               42  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fixed_variants = filtered_variants[filtered_variants['frequency'] > (1 - min_freq)].reset_index(drop=True)\n",
    "fixed_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "526f879c",
   "metadata": {},
   "source": [
    "Convert the remaining variants with a frequency > 50% into 'minor alleles' by reversing the reference and alternative columns and subtracting the allele frequency from 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "6337c970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "      <th>Accession</th>\n",
       "      <th>region</th>\n",
       "      <th>region_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8324</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>385</td>\n",
       "      <td>219</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>407.0</td>\n",
       "      <td>2.172830e-254</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>SRR13963679</td>\n",
       "      <td>E3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1260</td>\n",
       "      <td>507</td>\n",
       "      <td>78</td>\n",
       "      <td>77</td>\n",
       "      <td>0.058252</td>\n",
       "      <td>1339.0</td>\n",
       "      <td>4.275910e-17</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR12789644</td>\n",
       "      <td>E3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1278</td>\n",
       "      <td>552</td>\n",
       "      <td>74</td>\n",
       "      <td>73</td>\n",
       "      <td>0.054734</td>\n",
       "      <td>1352.0</td>\n",
       "      <td>4.224660e-30</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR12789651</td>\n",
       "      <td>E3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1316</td>\n",
       "      <td>527</td>\n",
       "      <td>86</td>\n",
       "      <td>84</td>\n",
       "      <td>0.061297</td>\n",
       "      <td>1403.0</td>\n",
       "      <td>2.045900e-35</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR12789645</td>\n",
       "      <td>E3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1374</td>\n",
       "      <td>574</td>\n",
       "      <td>77</td>\n",
       "      <td>75</td>\n",
       "      <td>0.052957</td>\n",
       "      <td>1454.0</td>\n",
       "      <td>4.548830e-16</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>SRR12789646</td>\n",
       "      <td>E3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  REF_RV  ALT_DP  ALT_RV  frequency   depth  \\\n",
       "0  8324   C   A     385     219      20      10   0.054054   407.0   \n",
       "1  8327   C   A    1260     507      78      77   0.058252  1339.0   \n",
       "2  8327   C   A    1278     552      74      73   0.054734  1352.0   \n",
       "3  8327   C   A    1316     527      86      84   0.061297  1403.0   \n",
       "4  8327   C   A    1374     574      77      75   0.052957  1454.0   \n",
       "\n",
       "         p_value REF_CODON REF_AA ALT_CODON ALT_AA    Accession region  \\\n",
       "0  2.172830e-254       CGT      R       AGT      S  SRR13963679     E3   \n",
       "1   4.275910e-17       CTT      L       ATT      I  SRR12789644     E3   \n",
       "2   4.224660e-30       CTT      L       ATT      I  SRR12789651     E3   \n",
       "3   2.045900e-35       CTT      L       ATT      I  SRR12789645     E3   \n",
       "4   4.548830e-16       CTT      L       ATT      I  SRR12789646     E3   \n",
       "\n",
       "   region_position  \n",
       "0                1  \n",
       "1                2  \n",
       "2                2  \n",
       "3                2  \n",
       "4                2  "
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the minor variants (anything that didn't end up in the fixed_variants dataframe)\n",
    "not_fixed_variants = filtered_variants[filtered_variants['frequency'] <= (1 - min_freq)].reset_index(drop=True)\n",
    "\n",
    "# Split dataframe based on frequency\n",
    "major_alleles = not_fixed_variants[not_fixed_variants['frequency'] > 0.5].copy()\n",
    "minor_alleles = not_fixed_variants[not_fixed_variants['frequency'] <= 0.5].copy()\n",
    "\n",
    "if len(major_alleles) > 0:\n",
    "    # Convert frequency for major alleles\n",
    "    major_alleles['frequency'] = 1 - major_alleles['frequency']\n",
    "    \n",
    "    # Swap REF and ALT columns\n",
    "    # First, identify all REF_ and ALT_ columns\n",
    "    ref_cols = [col for col in major_alleles.columns if col.startswith('REF_')]\n",
    "    alt_cols = [col for col in major_alleles.columns if col.startswith('ALT_')]\n",
    "    \n",
    "    # Create mapping for column swapping\n",
    "    swap_mapping = {}\n",
    "    for ref_col in ref_cols:\n",
    "        alt_col = ref_col.replace('REF_', 'ALT_')\n",
    "        if alt_col in alt_cols:\n",
    "            swap_mapping[ref_col] = alt_col\n",
    "            swap_mapping[alt_col] = ref_col\n",
    "    \n",
    "    # Also swap the basic REF and ALT columns\n",
    "    swap_mapping['REF'] = 'ALT'\n",
    "    swap_mapping['ALT'] = 'REF'\n",
    "    \n",
    "    # Perform the swapping\n",
    "    major_alleles = major_alleles.rename(columns=swap_mapping)\n",
    "    \n",
    "    # Reorder columns to match original dataframe\n",
    "    major_alleles = major_alleles[not_fixed_variants.columns]\n",
    "\n",
    "# Combine the dataframes\n",
    "joined_df = pd.concat([minor_alleles, major_alleles], ignore_index=True)\n",
    "\n",
    "# Sort by original index to maintain some order (optional)\n",
    "minor_variants_df = joined_df.sort_values('POS').reset_index(drop=True)\n",
    "minor_variants_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3462c691",
   "metadata": {},
   "source": [
    "Save each of these tables to the summary directory to be tracked by `git`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "0f8f4f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a 'summary' directory in results if it doesn't exist\n",
    "summary_dir = \"../../results/summary\"\n",
    "if not os.path.exists(summary_dir):\n",
    "    os.makedirs(summary_dir)\n",
    "# Save the minor variants to a CSV file\n",
    "minor_variants_csv_path = os.path.join(summary_dir, \"minor_variants.csv\")\n",
    "minor_variants_df.to_csv(minor_variants_csv_path, index=False)\n",
    "# Save the fixed variants to a CSV file\n",
    "fixed_variants_csv_path = os.path.join(summary_dir, \"fixed_variants.csv\")\n",
    "fixed_variants.to_csv(fixed_variants_csv_path, index=False) \n",
    "# Save the filtered variants to a CSV file\n",
    "filtered_variants_csv_path = os.path.join(summary_dir, \"all_variants.csv\")\n",
    "filtered_variants.to_csv(filtered_variants_csv_path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "intrahost",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
