{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66ee1de3",
   "metadata": {},
   "source": [
    "# Filter and save `ivar` variants\n",
    "\n",
    "I used `ivar` to identify variants relative to the sequence of the DMS library strain [MW473668](https://www.ncbi.nlm.nih.gov/nuccore/MW473668). In this notebook, I filter the variants to get only the coding mutations in the CHIKV E glycoprotein. I annotate these mutations with SRA metadata, sample coverage, and the amino acid in the library strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8505eb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "35dee6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_variants = pd.read_csv(\"../../results/variants/variants.ivar.csv\")\n",
    "sra_metadata = pd.read_csv(\"../../data/SRA_Runs.csv\")\n",
    "sitemap = pd.read_csv(\"../../../data/site_numbering_map.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6119a8c",
   "metadata": {},
   "source": [
    "## Process the variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e9ba11",
   "metadata": {},
   "source": [
    "Only keep the variants with a more than 100 reads at a site, a minimum frequency of 5%, a significant p-value < 0.05 and variants in the 'structural polyprotein' that encode the CHIKV E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "db6101bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the variants\n",
    "min_depth = 100\n",
    "min_freq = 0.05\n",
    "pvalue = 0.05\n",
    "feature = \"structural polyprotein:\"\n",
    "filtered_variants = raw_variants[\n",
    "    (raw_variants['GFF_FEATURE'] == feature) &\n",
    "    (raw_variants['TOTAL_DP'] >= min_depth) &\n",
    "    (raw_variants['ALT_FREQ'] >= min_freq) &\n",
    "    (raw_variants['PVAL'] <= pvalue)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bbbf301",
   "metadata": {},
   "source": [
    "Convert the numbering so it's relative to each region in the CHIKV E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "de95b972",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regions in the polyprotein\n",
    "polyprotein_numbering = {\n",
    "    \"C\": [1, 261],\n",
    "    \"E3\": [262, 325],\n",
    "    \"E2\": [326, 748],\n",
    "    \"6K\": [749, 809],\n",
    "    \"E1\": [810, 1248],\n",
    "}\n",
    "# Map the mature peptide region and position for each variant\n",
    "def map_peptide_and_position(pos_aa, polyprotein_numbering):\n",
    "    if pd.isna(pos_aa):\n",
    "        return None, None\n",
    "    \n",
    "    for peptide, (start, end) in polyprotein_numbering.items():\n",
    "        if start <= pos_aa <= end:\n",
    "            relative_pos = pos_aa - start + 1  # +1 for 1-based indexing\n",
    "            return peptide, relative_pos\n",
    "    \n",
    "    # If position doesn't fall within any range\n",
    "    return None, None\n",
    "# Apply the function to create both columns\n",
    "filtered_variants = filtered_variants.copy()\n",
    "filtered_variants[['peptide', 'peptide_position']] = filtered_variants['POS_AA'].apply(\n",
    "    lambda x: pd.Series(map_peptide_and_position(x, polyprotein_numbering))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "454ecc20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the useful columns from the metadata\n",
    "metadata_columns = ['Run', 'avgLength', 'LibraryStrategy', 'LibrarySelection', 'LibrarySource', 'LibraryLayout', 'Platform', 'Model', 'BioProject', 'ScientificName']\n",
    "sra_metadata = sra_metadata[metadata_columns]\n",
    "# Merge the filtered variants with SRA metadata\n",
    "annotated_variants = pd.merge(\n",
    "    filtered_variants,\n",
    "    sra_metadata,\n",
    "    left_on='Accession',\n",
    "    right_on='Run',\n",
    "    how='left'\n",
    ")\n",
    "# Add a column that makes a URL like to the BioProject\n",
    "annotated_variants['BioProject_URL'] = annotated_variants['BioProject'].apply(\n",
    "    lambda x: f\"https://www.ncbi.nlm.nih.gov/bioproject/?term={x}\" if pd.notna(x) else None\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b0fe17f",
   "metadata": {},
   "source": [
    "Only keep non-synonymous mutations in the subunits of the E protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4c48aa58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>avgLength</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>BioProject_URL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>308.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>308.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>308.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>7122</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>308.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9845</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1717</td>\n",
       "      <td>5524</td>\n",
       "      <td>0.755884</td>\n",
       "      <td>7308.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>...</td>\n",
       "      <td>308.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     POS REF ALT  REF_DP  ALT_DP  frequency   depth        p_value REF_CODON  \\\n",
       "7   9618   T   C       0     185   1.000000   185.0  5.638590e-131       GTG   \n",
       "8   9626   G   T       1     190   0.994764   191.0  1.848620e-132       GTG   \n",
       "9   9644   G   A       0     373   0.997326   374.0  1.421440e-260       GTA   \n",
       "13  9842   C   A       8    7122   0.997898  7137.0   0.000000e+00       CTG   \n",
       "14  9845   C   T    1717    5524   0.755884  7308.0   0.000000e+00       CAA   \n",
       "\n",
       "   REF_AA  ... avgLength LibraryStrategy LibrarySelection LibrarySource  \\\n",
       "7       V  ...     308.0             WGS       RANDOM PCR     VIRAL RNA   \n",
       "8       V  ...     308.0             WGS       RANDOM PCR     VIRAL RNA   \n",
       "9       V  ...     308.0             WGS       RANDOM PCR     VIRAL RNA   \n",
       "13      L  ...     308.0             WGS       RANDOM PCR     VIRAL RNA   \n",
       "14      Q  ...     308.0             WGS       RANDOM PCR     VIRAL RNA   \n",
       "\n",
       "    LibraryLayout  Platform           Model   BioProject     ScientificName  \\\n",
       "7          PAIRED  ILLUMINA  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "8          PAIRED  ILLUMINA  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "9          PAIRED  ILLUMINA  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "13         PAIRED  ILLUMINA  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "14         PAIRED  ILLUMINA  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "\n",
       "                                       BioProject_URL  \n",
       "7   https://www.ncbi.nlm.nih.gov/bioproject/?term=...  \n",
       "8   https://www.ncbi.nlm.nih.gov/bioproject/?term=...  \n",
       "9   https://www.ncbi.nlm.nih.gov/bioproject/?term=...  \n",
       "13  https://www.ncbi.nlm.nih.gov/bioproject/?term=...  \n",
       "14  https://www.ncbi.nlm.nih.gov/bioproject/?term=...  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop some columns from the dataframe\n",
    "columns_to_drop = [\"REGION\", \"GFF_FEATURE\", \"POS_AA\", \"PASS\", \"ALT_QUAL\", \"REF_QUAL\", \"Run\", \"REF_RV\", \"ALT_RV\"]\n",
    "annotated_variants.drop(columns=columns_to_drop, inplace=True)\n",
    "# Rename columns for clarity\n",
    "annotated_variants.rename(columns={\n",
    "    \"ALT_FREQ\": \"frequency\",\n",
    "    \"TOTAL_DP\": \"depth\",\n",
    "    \"PVAL\": \"p_value\",\n",
    "    \"peptide\": \"region\",\n",
    "    \"peptide_position\": \"region_position\"\n",
    "}, inplace=True)\n",
    "# Convert region_position to integer\n",
    "annotated_variants['region_position'] = annotated_variants['region_position'].astype('Int64')\n",
    "# Filter to keep only regions in the region list\n",
    "regions_to_keep = [\"E1\", \"E2\", \"6K\", \"E3\"]\n",
    "annotated_variants = annotated_variants[\n",
    "    annotated_variants['region'].isin(regions_to_keep)\n",
    "]\n",
    "# Only keep non-synonymous variants where REF_AA != ALT_AA\n",
    "annotated_variants = annotated_variants[\n",
    "    annotated_variants['REF_AA'] != annotated_variants['ALT_AA']\n",
    "]\n",
    "annotated_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1635db3b",
   "metadata": {},
   "source": [
    "Join the numbering from the DMS library to the variant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "44dca37b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>BioProject_URL</th>\n",
       "      <th>reference_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>sequential_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>368(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>371(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>377(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>7122</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>20(6K)</td>\n",
       "      <td>L</td>\n",
       "      <td>508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9845</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1717</td>\n",
       "      <td>5524</td>\n",
       "      <td>0.755884</td>\n",
       "      <td>7308.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>...</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>21(6K)</td>\n",
       "      <td>Q</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  ALT_DP  frequency   depth        p_value REF_CODON  \\\n",
       "0  9618   T   C       0     185   1.000000   185.0  5.638590e-131       GTG   \n",
       "1  9626   G   T       1     190   0.994764   191.0  1.848620e-132       GTG   \n",
       "2  9644   G   A       0     373   0.997326   374.0  1.421440e-260       GTA   \n",
       "3  9842   C   A       8    7122   0.997898  7137.0   0.000000e+00       CTG   \n",
       "4  9845   C   T    1717    5524   0.755884  7308.0   0.000000e+00       CAA   \n",
       "\n",
       "  REF_AA  ... LibrarySource LibraryLayout  Platform           Model  \\\n",
       "0      V  ...     VIRAL RNA        PAIRED  ILLUMINA  Illumina MiSeq   \n",
       "1      V  ...     VIRAL RNA        PAIRED  ILLUMINA  Illumina MiSeq   \n",
       "2      V  ...     VIRAL RNA        PAIRED  ILLUMINA  Illumina MiSeq   \n",
       "3      L  ...     VIRAL RNA        PAIRED  ILLUMINA  Illumina MiSeq   \n",
       "4      Q  ...     VIRAL RNA        PAIRED  ILLUMINA  Illumina MiSeq   \n",
       "\n",
       "    BioProject     ScientificName  \\\n",
       "0  PRJNA294670  Chikungunya virus   \n",
       "1  PRJNA294670  Chikungunya virus   \n",
       "2  PRJNA294670  Chikungunya virus   \n",
       "3  PRJNA294670  Chikungunya virus   \n",
       "4  PRJNA294670  Chikungunya virus   \n",
       "\n",
       "                                      BioProject_URL reference_site wildtype  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/bioproject/?term=...        368(E2)        V   \n",
       "1  https://www.ncbi.nlm.nih.gov/bioproject/?term=...        371(E2)        V   \n",
       "2  https://www.ncbi.nlm.nih.gov/bioproject/?term=...        377(E2)        V   \n",
       "3  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         20(6K)        L   \n",
       "4  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         21(6K)        Q   \n",
       "\n",
       "  sequential_site  \n",
       "0             433  \n",
       "1             436  \n",
       "2             442  \n",
       "3             508  \n",
       "4             509  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a column called `reference_site` by joining the region and region_position column like region_position(region)\n",
    "annotated_variants['reference_site'] = annotated_variants.apply(\n",
    "    lambda row: f\"{row['region_position']}({row['region']})\" if pd.notna(row['region_position']) else None,\n",
    "    axis=1\n",
    ")\n",
    "# Join with the site_numbering_map by reference_site\n",
    "annotated_variants = pd.merge(\n",
    "    annotated_variants,\n",
    "    sitemap[['reference_site', 'wildtype', 'sequential_site']],\n",
    "    on='reference_site',\n",
    "    how='left'\n",
    ")\n",
    "annotated_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97cf5262",
   "metadata": {},
   "source": [
    "Calculate read coverage from the depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "030780c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>Model</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>BioProject_URL</th>\n",
       "      <th>reference_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>sample_mean_read_depth</th>\n",
       "      <th>sample_percent_coverage</th>\n",
       "      <th>sample_percent_100X_coverage_breadth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>368(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>433</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>371(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>436</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>377(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>442</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>7122</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>20(6K)</td>\n",
       "      <td>L</td>\n",
       "      <td>508</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9845</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1717</td>\n",
       "      <td>5524</td>\n",
       "      <td>0.755884</td>\n",
       "      <td>7308.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>21(6K)</td>\n",
       "      <td>Q</td>\n",
       "      <td>509</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  ALT_DP  frequency   depth        p_value REF_CODON  \\\n",
       "0  9618   T   C       0     185   1.000000   185.0  5.638590e-131       GTG   \n",
       "1  9626   G   T       1     190   0.994764   191.0  1.848620e-132       GTG   \n",
       "2  9644   G   A       0     373   0.997326   374.0  1.421440e-260       GTA   \n",
       "3  9842   C   A       8    7122   0.997898  7137.0   0.000000e+00       CTG   \n",
       "4  9845   C   T    1717    5524   0.755884  7308.0   0.000000e+00       CAA   \n",
       "\n",
       "  REF_AA  ...           Model   BioProject     ScientificName  \\\n",
       "0      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "1      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "2      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "3      L  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "4      Q  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "\n",
       "                                      BioProject_URL  reference_site  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         368(E2)   \n",
       "1  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         371(E2)   \n",
       "2  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         377(E2)   \n",
       "3  https://www.ncbi.nlm.nih.gov/bioproject/?term=...          20(6K)   \n",
       "4  https://www.ncbi.nlm.nih.gov/bioproject/?term=...          21(6K)   \n",
       "\n",
       "   wildtype sequential_site sample_mean_read_depth sample_percent_coverage  \\\n",
       "0         V             433            1566.448549                0.998332   \n",
       "1         V             436            1566.448549                0.998332   \n",
       "2         V             442            1566.448549                0.998332   \n",
       "3         L             508            1566.448549                0.998332   \n",
       "4         Q             509            1566.448549                0.998332   \n",
       "\n",
       "  sample_percent_100X_coverage_breadth  \n",
       "0                             0.574967  \n",
       "1                             0.574967  \n",
       "2                             0.574967  \n",
       "3                             0.574967  \n",
       "4                             0.574967  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in the read coverage over each position in the genome\n",
    "accessions_to_keep = set(annotated_variants.Accession.to_list())\n",
    "depth_df = pd.read_csv(\"../../results/depth/merged.depth\", sep=\"\\t\")\n",
    "# Filter to only the relevant accessions\n",
    "depth_df = depth_df[depth_df['Accession'].isin(accessions_to_keep)]\n",
    "# Calculate the % coverage and mean read depth per sample\n",
    "genome_length = 11992\n",
    "coverage_summary_df = depth_df.groupby('Accession')['DP'].agg([\n",
    "    ('sample_mean_read_depth', 'mean'),\n",
    "    ('sample_coverage', lambda x: (x > 0).sum()),\n",
    "    ('sample_100x_coverage', lambda x: (x > 100).sum())\n",
    "]).reset_index()\n",
    "coverage_summary_df = coverage_summary_df.assign(\n",
    "    sample_percent_coverage=lambda x: x['sample_coverage'] / genome_length,\n",
    "    sample_percent_100X_coverage_breadth=lambda x: x['sample_100x_coverage'] / genome_length\n",
    ")\n",
    "coverage_summary_df.drop(columns=['sample_coverage', 'sample_100x_coverage'], inplace=True)\n",
    "# Join the coverage summary to the annotated variants by Accession\n",
    "annotated_variants = pd.merge(\n",
    "    annotated_variants,\n",
    "    coverage_summary_df,\n",
    "    on='Accession',\n",
    "    how='left'\n",
    ")\n",
    "annotated_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfbcef4c",
   "metadata": {},
   "source": [
    "Get the 'fixed' variants. These are variants with a frequency that's greater than `1 - minimum frequency` such that they cannot be converted to minor alleles with a frequency > our `minimum frequency`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "03c34181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>Model</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>BioProject_URL</th>\n",
       "      <th>reference_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>sample_mean_read_depth</th>\n",
       "      <th>sample_percent_coverage</th>\n",
       "      <th>sample_percent_100X_coverage_breadth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9618</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>185.0</td>\n",
       "      <td>5.638590e-131</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>368(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>433</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9626</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>0.994764</td>\n",
       "      <td>191.0</td>\n",
       "      <td>1.848620e-132</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>371(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>436</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9644</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>373</td>\n",
       "      <td>0.997326</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1.421440e-260</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>377(E2)</td>\n",
       "      <td>V</td>\n",
       "      <td>442</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>7122</td>\n",
       "      <td>0.997898</td>\n",
       "      <td>7137.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>20(6K)</td>\n",
       "      <td>L</td>\n",
       "      <td>508</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9909</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>28</td>\n",
       "      <td>8577</td>\n",
       "      <td>0.995705</td>\n",
       "      <td>8614.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PRJNA294670</td>\n",
       "      <td>Chikungunya virus</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>42(6K)</td>\n",
       "      <td>F</td>\n",
       "      <td>530</td>\n",
       "      <td>1566.448549</td>\n",
       "      <td>0.998332</td>\n",
       "      <td>0.574967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  ALT_DP  frequency   depth        p_value REF_CODON  \\\n",
       "0  9618   T   C       0     185   1.000000   185.0  5.638590e-131       GTG   \n",
       "1  9626   G   T       1     190   0.994764   191.0  1.848620e-132       GTG   \n",
       "2  9644   G   A       0     373   0.997326   374.0  1.421440e-260       GTA   \n",
       "3  9842   C   A       8    7122   0.997898  7137.0   0.000000e+00       CTG   \n",
       "4  9909   T   G      28    8577   0.995705  8614.0   0.000000e+00       TTT   \n",
       "\n",
       "  REF_AA  ...           Model   BioProject     ScientificName  \\\n",
       "0      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "1      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "2      V  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "3      L  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "4      F  ...  Illumina MiSeq  PRJNA294670  Chikungunya virus   \n",
       "\n",
       "                                      BioProject_URL  reference_site  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         368(E2)   \n",
       "1  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         371(E2)   \n",
       "2  https://www.ncbi.nlm.nih.gov/bioproject/?term=...         377(E2)   \n",
       "3  https://www.ncbi.nlm.nih.gov/bioproject/?term=...          20(6K)   \n",
       "4  https://www.ncbi.nlm.nih.gov/bioproject/?term=...          42(6K)   \n",
       "\n",
       "   wildtype sequential_site sample_mean_read_depth sample_percent_coverage  \\\n",
       "0         V             433            1566.448549                0.998332   \n",
       "1         V             436            1566.448549                0.998332   \n",
       "2         V             442            1566.448549                0.998332   \n",
       "3         L             508            1566.448549                0.998332   \n",
       "4         F             530            1566.448549                0.998332   \n",
       "\n",
       "  sample_percent_100X_coverage_breadth  \n",
       "0                             0.574967  \n",
       "1                             0.574967  \n",
       "2                             0.574967  \n",
       "3                             0.574967  \n",
       "4                             0.574967  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fixed_variants = annotated_variants[annotated_variants['frequency'] > (1 - min_freq)].reset_index(drop=True)\n",
    "fixed_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "526f879c",
   "metadata": {},
   "source": [
    "Convert the remaining variants with a frequency > 50% into 'minor alleles' by reversing the reference and alternative columns and subtracting the allele frequency from 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "6337c970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>frequency</th>\n",
       "      <th>depth</th>\n",
       "      <th>p_value</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>Model</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>BioProject_URL</th>\n",
       "      <th>reference_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>sample_mean_read_depth</th>\n",
       "      <th>sample_percent_coverage</th>\n",
       "      <th>sample_percent_100X_coverage_breadth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8324</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>385</td>\n",
       "      <td>20</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>407.0</td>\n",
       "      <td>2.172830e-254</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>PRJNA714555</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>1(E3)</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>3886.443045</td>\n",
       "      <td>0.091728</td>\n",
       "      <td>0.065460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1260</td>\n",
       "      <td>78</td>\n",
       "      <td>0.058252</td>\n",
       "      <td>1339.0</td>\n",
       "      <td>4.275910e-17</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>PRJNA667927</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>2(E3)</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>663.712141</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.993245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1278</td>\n",
       "      <td>74</td>\n",
       "      <td>0.054734</td>\n",
       "      <td>1352.0</td>\n",
       "      <td>4.224660e-30</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>PRJNA667927</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>2(E3)</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>640.163192</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.993579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1316</td>\n",
       "      <td>86</td>\n",
       "      <td>0.061297</td>\n",
       "      <td>1403.0</td>\n",
       "      <td>2.045900e-35</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>PRJNA667927</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>2(E3)</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>690.586558</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8327</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1374</td>\n",
       "      <td>77</td>\n",
       "      <td>0.052957</td>\n",
       "      <td>1454.0</td>\n",
       "      <td>4.548830e-16</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>PRJNA667927</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>2(E3)</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>709.473399</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.996998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT  REF_DP  ALT_DP  frequency   depth        p_value REF_CODON  \\\n",
       "0  8324   C   A     385      20   0.054054   407.0  2.172830e-254       CGT   \n",
       "1  8327   C   A    1260      78   0.058252  1339.0   4.275910e-17       CTT   \n",
       "2  8327   C   A    1278      74   0.054734  1352.0   4.224660e-30       CTT   \n",
       "3  8327   C   A    1316      86   0.061297  1403.0   2.045900e-35       CTT   \n",
       "4  8327   C   A    1374      77   0.052957  1454.0   4.548830e-16       CTT   \n",
       "\n",
       "  REF_AA  ...                  Model   BioProject ScientificName  \\\n",
       "0      R  ...    Illumina HiSeq 4000  PRJNA714555   Homo sapiens   \n",
       "1      L  ...  Illumina NovaSeq 6000  PRJNA667927   Homo sapiens   \n",
       "2      L  ...  Illumina NovaSeq 6000  PRJNA667927   Homo sapiens   \n",
       "3      L  ...  Illumina NovaSeq 6000  PRJNA667927   Homo sapiens   \n",
       "4      L  ...  Illumina NovaSeq 6000  PRJNA667927   Homo sapiens   \n",
       "\n",
       "                                      BioProject_URL  reference_site  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/bioproject/?term=...           1(E3)   \n",
       "1  https://www.ncbi.nlm.nih.gov/bioproject/?term=...           2(E3)   \n",
       "2  https://www.ncbi.nlm.nih.gov/bioproject/?term=...           2(E3)   \n",
       "3  https://www.ncbi.nlm.nih.gov/bioproject/?term=...           2(E3)   \n",
       "4  https://www.ncbi.nlm.nih.gov/bioproject/?term=...           2(E3)   \n",
       "\n",
       "   wildtype sequential_site sample_mean_read_depth sample_percent_coverage  \\\n",
       "0         S               2            3886.443045                0.091728   \n",
       "1         L               3             663.712141                1.000000   \n",
       "2         L               3             640.163192                1.000000   \n",
       "3         L               3             690.586558                1.000000   \n",
       "4         L               3             709.473399                1.000000   \n",
       "\n",
       "  sample_percent_100X_coverage_breadth  \n",
       "0                             0.065460  \n",
       "1                             0.993245  \n",
       "2                             0.993579  \n",
       "3                             0.995247  \n",
       "4                             0.996998  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the minor variants (anything that didn't end up in the fixed_variants dataframe)\n",
    "not_fixed_variants = annotated_variants[annotated_variants['frequency'] <= (1 - min_freq)].reset_index(drop=True)\n",
    "\n",
    "# Split dataframe based on frequency\n",
    "major_alleles = not_fixed_variants[not_fixed_variants['frequency'] > 0.5].copy()\n",
    "minor_alleles = not_fixed_variants[not_fixed_variants['frequency'] <= 0.5].copy()\n",
    "\n",
    "if len(major_alleles) > 0:\n",
    "    # Convert frequency for major alleles\n",
    "    major_alleles['frequency'] = 1 - major_alleles['frequency']\n",
    "    \n",
    "    # Swap REF and ALT columns\n",
    "    # First, identify all REF_ and ALT_ columns\n",
    "    ref_cols = [col for col in major_alleles.columns if col.startswith('REF_')]\n",
    "    alt_cols = [col for col in major_alleles.columns if col.startswith('ALT_')]\n",
    "    \n",
    "    # Create mapping for column swapping\n",
    "    swap_mapping = {}\n",
    "    for ref_col in ref_cols:\n",
    "        alt_col = ref_col.replace('REF_', 'ALT_')\n",
    "        if alt_col in alt_cols:\n",
    "            swap_mapping[ref_col] = alt_col\n",
    "            swap_mapping[alt_col] = ref_col\n",
    "    \n",
    "    # Also swap the basic REF and ALT columns\n",
    "    swap_mapping['REF'] = 'ALT'\n",
    "    swap_mapping['ALT'] = 'REF'\n",
    "    \n",
    "    # Perform the swapping\n",
    "    major_alleles = major_alleles.rename(columns=swap_mapping)\n",
    "    \n",
    "    # Reorder columns to match original dataframe\n",
    "    major_alleles = major_alleles[not_fixed_variants.columns]\n",
    "\n",
    "# Combine the dataframes\n",
    "joined_df = pd.concat([minor_alleles, major_alleles], ignore_index=True)\n",
    "\n",
    "# Sort by original index to maintain some order (optional)\n",
    "minor_variants_df = joined_df.sort_values('POS').reset_index(drop=True)\n",
    "minor_variants_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af5649c3",
   "metadata": {},
   "source": [
    "Rename the columns to something easier to understand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "974ec5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_dictionary = {\n",
    "    'POS': 'nucleotide_position',\n",
    "    'REF': 'reference_nucleotide',\n",
    "    'ALT': 'variant_nucleotide',\n",
    "    'REF_DP': 'reference_depth',\n",
    "    'ALT_DP': 'variant_depth',\n",
    "    'frequency': 'variant_frequency',\n",
    "    'depth': 'read_depth',\n",
    "    'REF_CODON': 'reference_codon',\n",
    "    'REF_AA': 'reference_amino_acid',\n",
    "    'ALT_CODON': 'variant_codon',\n",
    "    'ALT_AA': 'variant_amino_acid',\n",
    "    'avgLength': 'avg_read_length',\n",
    "    'LibraryStrategy': 'library_strategy',\n",
    "    'LibrarySelection': 'library_selection',\n",
    "    'LibrarySource': 'library_source',\n",
    "    'LibraryLayout': 'library_layout',\n",
    "    'Platform': 'sequencing_platform',\n",
    "    'Model': 'sequencing_model',\n",
    "    'ScientificName': 'library_organism',\n",
    "    'wildtype': 'library_amino_acid'\n",
    " }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3462c691",
   "metadata": {},
   "source": [
    "Save each of these tables to the summary directory to be tracked by `git`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "0f8f4f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a 'summary' directory in results if it doesn't exist\n",
    "summary_dir = \"../../results/summary\"\n",
    "if not os.path.exists(summary_dir):\n",
    "    os.makedirs(summary_dir)\n",
    "# Save the minor variants to a CSV file\n",
    "minor_variants_csv_path = os.path.join(summary_dir, \"minor_variants.csv\")\n",
    "minor_variants_df.rename(columns=rename_dictionary).to_csv(minor_variants_csv_path, index=False)\n",
    "# Save the fixed variants to a CSV file\n",
    "fixed_variants_csv_path = os.path.join(summary_dir, \"fixed_variants.csv\")\n",
    "fixed_variants.rename(columns=rename_dictionary).to_csv(fixed_variants_csv_path, index=False) \n",
    "# Save the filtered variants to a CSV file\n",
    "filtered_variants_csv_path = os.path.join(summary_dir, \"all_variants.csv\")\n",
    "annotated_variants.rename(columns=rename_dictionary).to_csv(filtered_variants_csv_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12474d30",
   "metadata": {},
   "source": [
    "## Basic variant information\n",
    "\n",
    "How many variants are there, how many BioProjects and BioSamples are they coming from?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "21ec3ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "minor_variants_df[\"amino_acid_variant\"] = (minor_variants_df['REF_AA'] + \n",
    "                                          minor_variants_df['sequential_site'].astype(str) + \n",
    "                                          minor_variants_df['ALT_AA'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "38ec649f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1075 unique minor variants in the dataset.\n"
     ]
    }
   ],
   "source": [
    "print(f\"There are {len(set(minor_variants_df.amino_acid_variant.to_list()))} unique minor variants in the dataset.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "3d5a89e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These were identified in 1173 unique SRA runs.\n"
     ]
    }
   ],
   "source": [
    "print(f\"These were identified in {len(set(minor_variants_df.Accession.to_list()))} unique SRA runs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "edd40465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "From 75 unique BioProjects.\n"
     ]
    }
   ],
   "source": [
    "print(f\"From {len(set(minor_variants_df.BioProject.to_list()))} unique BioProjects.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "intrahost",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
